File: convex/knowledge.ts
import { query, mutation } from './_generated/server';
import { v } from 'convex/values';
import { getAuthUserId } from '@convex-dev/auth/server';

// Helper function to get authenticated user
async function getAuthenticatedUser(ctx: any) {
	const userId = await getAuthUserId(ctx);
	if (!userId) {
		throw new Error('Not authenticated');
	}
	return userId;
}

// Search knowledge base
export const searchKnowledge = query({
	args: {
		searchTerm: v.string(),
		category: v.optional(
			v.union(
				v.literal('case_study'),
				v.literal('solution_template'),
				v.literal('pricing_model'),
				v.literal('team_bio'),
				v.literal('company_overview'),
				v.literal('technical_spec')
			)
		),
		industry: v.optional(v.string()),
	},
	handler: async (ctx, args) => {
		await getAuthenticatedUser(ctx);

		const searchQuery = ctx.db
			.query('knowledgeBase')
			.withSearchIndex('search_content', (q) => {
				let query = q.search('content', args.searchTerm);
				if (args.category) {
					query = query.eq('category', args.category);
				}
				if (args.industry) {
					query = query.eq('industry', args.industry);
				}
				return query.eq('isApproved', true);
			});

		const results = await searchQuery.take(20);

		// Get creator details for each result
		const resultsWithCreators = await Promise.all(
			results.map(async (item) => {
				const creator = await ctx.db.get(item.createdBy);
				return {
					...item,
					creator: creator
						? { name: creator.name, email: creator.email }
						: null,
				};
			})
		);

		return resultsWithCreators;
	},
});

// List knowledge base items by category
export const listKnowledgeByCategory = query({
	args: {
		category: v.union(
			v.literal('case_study'),
			v.literal('solution_template'),
			v.literal('pricing_model'),
			v.literal('team_bio'),
			v.literal('company_overview'),
			v.literal('technical_spec')
		),
	},
	handler: async (ctx, args) => {
		await getAuthenticatedUser(ctx);

		const items = await ctx.db
			.query('knowledgeBase')
			.withIndex('by_category', (q) => q.eq('category', args.category))
			.filter((q) => q.eq(q.field('isApproved'), true))
			.collect();

		// Get creator details for each item
		const itemsWithCreators = await Promise.all(
			items.map(async (item) => {
				const creator = await ctx.db.get(item.createdBy);
				return {
					...item,
					creator: creator
						? { name: creator.name, email: creator.email }
						: null,
				};
			})
		);

		return itemsWithCreators;
	},
});

// Get knowledge base item
export const getKnowledgeItem = query({
	args: { itemId: v.id('knowledgeBase') },
	handler: async (ctx, args) => {
		await getAuthenticatedUser(ctx);

		const item = await ctx.db.get(args.itemId);
		if (!item) {
			throw new Error('Knowledge item not found');
		}

		const creator = await ctx.db.get(item.createdBy);

		return {
			...item,
			creator: creator ? { name: creator.name, email: creator.email } : null,
		};
	},
});

// Create knowledge base item
export const createKnowledgeItem = mutation({
	args: {
		title: v.string(),
		content: v.string(),
		category: v.union(
			v.literal('case_study'),
			v.literal('solution_template'),
			v.literal('pricing_model'),
			v.literal('team_bio'),
			v.literal('company_overview'),
			v.literal('technical_spec')
		),
		tags: v.array(v.string()),
		industry: v.optional(v.string()),
	},
	handler: async (ctx, args) => {
		const userId = await getAuthenticatedUser(ctx);

		const itemId = await ctx.db.insert('knowledgeBase', {
			title: args.title,
			content: args.content,
			category: args.category,
			tags: args.tags,
			industry: args.industry,
			createdBy: userId,
			usageCount: 0,
			isApproved: false, // Requires approval
		});

		return itemId;
	},
});

// Approve knowledge base item
export const approveKnowledgeItem = mutation({
	args: { itemId: v.id('knowledgeBase') },
	handler: async (ctx, args) => {
		const userId = await getAuthenticatedUser(ctx);

		// Check if user has approval permissions
		const userProfile = await ctx.db
			.query('userProfiles')
			.withIndex('by_user', (q) => q.eq('userId', userId))
			.first();

		if (!userProfile || !userProfile.permissions.includes('approve')) {
			throw new Error('Insufficient permissions to approve knowledge items');
		}

		await ctx.db.patch(args.itemId, {
			isApproved: true,
		});

		return args.itemId;
	},
});

// Increment usage count
export const incrementUsageCount = mutation({
	args: { itemId: v.id('knowledgeBase') },
	handler: async (ctx, args) => {
		await getAuthenticatedUser(ctx);

		const item = await ctx.db.get(args.itemId);
		if (!item) {
			throw new Error('Knowledge item not found');
		}

		await ctx.db.patch(args.itemId, {
			usageCount: item.usageCount + 1,
		});

		return args.itemId;
	},
});

File: backend/app/routers/knowledge.py
from typing import List, Optional
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import or_, and_
from ..database import get_db
from ..models import KnowledgeBase, User, UserProfile
from ..schemas import KnowledgeBaseCreate, KnowledgeBaseResponse
from ..auth import get_current_active_user
import json

router = APIRouter(prefix="/knowledge", tags=["knowledge"])

@router.post("/", response_model=KnowledgeBaseResponse)
def create_knowledge_item(
    knowledge: KnowledgeBaseCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    db_knowledge = KnowledgeBase(
        title=knowledge.title,
        content=knowledge.content,
        category=knowledge.category,
        tags=json.dumps(knowledge.tags),
        industry=knowledge.industry,
        created_by=current_user.id,
        is_approved=False  # Requires approval
    )
    db.add(db_knowledge)
    db.commit()
    db.refresh(db_knowledge)
    return db_knowledge

@router.get("/", response_model=List[KnowledgeBaseResponse])
def list_knowledge_items(
    skip: int = 0,
    limit: int = 100,
    category: Optional[str] = None,
    industry: Optional[str] = None,
    approved_only: bool = True,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    query = db.query(KnowledgeBase)
    
    if approved_only:
        query = query.filter(KnowledgeBase.is_approved == True)
    if category:
        query = query.filter(KnowledgeBase.category == category)
    if industry:
        query = query.filter(KnowledgeBase.industry == industry)
    
    knowledge_items = query.offset(skip).limit(limit).all()
    return knowledge_items

@router.get("/search", response_model=List[KnowledgeBaseResponse])
def search_knowledge(
    q: str,
    category: Optional[str] = None,
    industry: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    query = db.query(KnowledgeBase).filter(
        and_(
            KnowledgeBase.is_approved == True,
            or_(
                KnowledgeBase.title.ilike(f"%{q}%"),
                KnowledgeBase.content.ilike(f"%{q}%")
            )
        )
    )
    
    if category:
        query = query.filter(KnowledgeBase.category == category)
    if industry:
        query = query.filter(KnowledgeBase.industry == industry)
    
    results = query.limit(20).all()
    return results

@router.put("/{knowledge_id}/approve")
def approve_knowledge_item(
    knowledge_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    # Check if user has approval permissions
    user_profile = db.query(UserProfile).filter(UserProfile.user_id == current_user.id).first()
    if not user_profile:
        raise HTTPException(status_code=403, detail="User profile not found")
    
    permissions = json.loads(user_profile.permissions) if user_profile.permissions else []
    if "approve" not in permissions:
        raise HTTPException(status_code=403, detail="Insufficient permissions")
    
    knowledge_item = db.query(KnowledgeBase).filter(KnowledgeBase.id == knowledge_id).first()
    if not knowledge_item:
        raise HTTPException(status_code=404, detail="Knowledge item not found")
    
    knowledge_item.is_approved = True
    db.commit()
    
    return {"message": "Knowledge item approved"}

@router.put("/{knowledge_id}/increment-usage")
def increment_usage_count(
    knowledge_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
):
    knowledge_item = db.query(KnowledgeBase).filter(KnowledgeBase.id == knowledge_id).first()
    if not knowledge_item:
        raise HTTPException(status_code=404, detail="Knowledge item not found")
    
    knowledge_item.usage_count += 1
    db.commit()
    
    return {"message": "Usage count incremented"}
